### PR TITLE
[Feature] 8886 custom fields access too restricted

### PR DESCRIPTION
**Pull request by hschink**
[`* `#8886`Custom field access too restricted`](https://www.openproject.org/work_packages/8886)

### CODE REVIEW

File: app/models/work_package_custom_field.rb [hunk 1]
```diff
@@ -32,6 +32,15 @@ class WorkPackageCustomField < CustomField
   has_and_belongs_to_many :types, :join_table => "#{table_name_prefix}custom_fields_types#{table_name_suffix}", :foreign_key => "custom_field_id"
   has_many :work_packages, :through => :work_package_custom_values
 
+  scope :visible_by_user, lambda { |user|
+    unless user.admin?
+      joins("LEFT OUTER JOIN custom_fields_projects AS cfp ON (custom_fields.id = cfp.custom_field_id) " \
+            "LEFT OUTER JOIN projects AS p ON (cfp.project_id = p.id) " \
+            "LEFT OUTER JOIN members AS m ON (p.id = m.project_id)")
+      .where("custom_fields.is_for_all = #{ActiveRecord::Base.connection.quoted_true} OR m.user_id = ?", user.id)
```
Comment (ulferts) [inline app/models/work_package_custom_field.rb]: Public projects are not taken into account I guess. A user being a non_member will not have such a relationship but might still be allowed to access the project.

Comment (hschink) [inline app/models/work_package_custom_field.rb]: @ulferts Thanks for the hint! I fixed that in the following commits.

File: app/controllers/api/v2/custom_fields_controller.rb [hunk 2]
```diff
@@ -63,8 +65,8 @@ def show
 
       protected
 
-      def require_permissions
-        deny_access unless User.current.allowed_to? :edit_project, nil, :global => true
+      def require_authentication
+        deny_access if User.current.anonymous?
```
Comment (ulferts) [inline app/controllers/api/v2/custom_fields_controller.rb]: I might be contradicting myself here but I guess that Anonymous should be granted access as long as the require_login setting doesn't prevent it (should already be enforced in a different before_filter. 

The reason I am changing my mind here are the public projects which are OK for Anonymous to see.

So I guess this before filter here can be removed.

Comment (hschink) [inline app/controllers/api/v2/custom_fields_controller.rb]: @ulferts Done.

PS.: I need to rebase the branch because of a change in the core.

**Pull request merged**